### DISCUSSION TITLE
Hide and unhide objects from their ECinstanceId

### DISCUSSION THREAD

**Discussion by tusharsable** [iTwin/itwinjs-core#2526] [Q&A]
We want to hide objects when the user selects a specific category of object and presses a key . and unhide them with toggle. If we could get a sample code to hide object if we have its ECInstanceid from model

**Comment by tusharsable** [comment]
I Found Sample code for hiding [here](https://www.itwinjs.org/sample-showcase/?group=Viewer+Features&sample=emphasize-elements-sample&imodel=Retail+Building+Sample)

**Reply by tusharsable** [reply]
const vp = IModelApp.viewManager.selectedView!
    const emph = EmphasizeElements.getOrCreate(vp);
    emph.hideElements(SelectionEcInstanceId,vp,true);

I tried using the above code for hiding selected elements which fulfill a criteria but the element doesn't hide.
Above code is executed on button click event

**Reply by tusharsable** [reply]
@pmconne Is the above code correct for the said purpose

**Reply by bbastings** [reply]
Show how you got SelectionEcInstanceId.

**Reply by tusharsable** [reply]
```
 const manekinListener=async (prev:SelectionSetEvent)=>{
  const query = `Select ECInstanceId From RevitDynamic.Manikin`;
  const imodel = UiFramework.getIModelConnection()!;
    const result = await imodel.query(query)
    const Id = [...prev.set.elements];
    console.log(result)
    for await (const row of result){
      console.log(row);
     if(Id[0]===row.id){
      setSelectionEcInstanceId(Id[0]);
      break;
      }
    }
```

@bbastings  where SelectionEcInstance id is a useState hook and manekinListener is a listener added for selection in viewer

**Comment by pmconne** [comment]
```
     if(Id[0]===row.id){
      `setSelectionEcInstanceId(Id[0]);`
      break;
```

Why is this line in backticks? That makes it just a string, not an executable statement.
Did you verify your `console.log` is producing expected results?
Did you verify contents of SelectionEcInstanceId in debugger?

**Reply by bbastings** [reply]
What is the size of the Id array? Why are you only trying to hide Id[0]? If for example an assembly is was added to the selection set (assembly selection scope), the "selection" will hilite multiple elements, Id[0] could be the assembly parent, which may not have any graphics. You'd need to hide the entire assembly, not just the parent...

Seems like you should be checking if any of the selected elements (Id array) are of interest, not querying all possible ids of interest and comparing those to the selection...

**Reply by tusharsable** [reply]
When I select the object of interest The size of Id is just one. which I compare.

**Reply by tusharsable** [reply]
Yes the console.log  gives the instance Id of all objects of manikin category {id: '0x20000000324'}

**Reply by tusharsable** [reply]
The backticks are a mistake which I made when posting code here

**Reply by tusharsable** [reply]
Yes I would try the approach of querying

**Reply by tusharsable** [reply]
@bbastings  The Item I Want to Hide Individually is a part of a group in tree. I can't hide the Item from the tree. The whole group can be hidden though. Also when I select individual Item the whole group gets selected

**Reply by bbastings** [reply]
It sounds like you have an element that is part of an assembly and the selection scope is set to "assembly" or "top assembly". So when you pick any member of an assembly, the assembly parent is selected and all members of the assembly are hilited. If the selectionn scope was set to "element", you likely get the behavior you want (i.e. just the element you pick is selected/hilited).

https://www.itwinjs.org/learning/presentation/unified-selection/terminology/
https://www.itwinjs.org/reference/ui-framework/frontstage/selectionscope/

note: There is a status bar item for letting the user specify the selection scope...not sure what viewer you are using...

**Reply by tusharsable** [reply]
@bbastings  I Changed the scope to Element, Now I could Select the Manikin individually, But Can't Hide it still using the viewer menu button to hide elements

**Reply by tusharsable** [reply]
Hi @bbastings I can't hide the element using the default
![image](https://user-images.githubusercontent.com/66006244/139004270-e64c37a8-6d0d-4d3a-b576-406285c34a96.png)
button available on viewer using element selection mode. What could be the reason for this?

**Reply by bbastings** [reply]
Assuming it's actually a geometric element with displayable geometry I can't think of a reason. You can see the element selected/hilited?

**Reply by tusharsable** [reply]
Hi @bbastings ,Yes the Item gets selected and highlighted when scope changed to element. I also observed that they get grouped when hidden . but geometry is still visible.

**Reply by bbastings** [reply]
> I also observed that they get grouped when hidden

Not sure what this means...possible to share model? Any chance you have duplicate/coincident elements?
